File: Sources/Terminal/AngouriMath.Terminal.Lib/FSharpInteractive.fs
module AngouriMath.Terminal.Lib.FSharpInteractive

open AngouriMath.Core
open Consts
open System.Text.Json
open System.Text.Json.Serialization
open Microsoft.DotNet.Interactive
open Microsoft.DotNet.Interactive.Events
open Microsoft.DotNet.Interactive.FSharp
open Microsoft.DotNet.Interactive.Commands
open Microsoft.DotNet.Interactive.Formatting
open System
open AngouriMath
open Plotly.NET
open AngouriMath.Terminal.Lib.Consts
open AssemblyLoadBuilder
open AngouriMath.FSharp.Core
open AngouriMath.InteractiveExtension

let options = JsonSerializerOptions ()
JsonFSharpConverter () |> options.Converters.Add

let private objectEncode (o : obj) =
    match o with
    | :? ILatexizeable as latexizeable -> 
        let toSerialize = LatexSuccess (latexizeable.Latexize (), string o)
        EncodingLatexPrefix + JsonSerializer.Serialize(toSerialize, options)
    | _ ->
        let toSerialize = PlainTextSuccess (string o)
        EncodingPlainPrefix + JsonSerializer.Serialize(toSerialize, options)


let private objectDecode (s : string Option) =
    match s with
    | None -> VoidSuccess
    | Some plain when plain.StartsWith EncodingPlainPrefix ->
        JsonSerializer.Deserialize<ExecutionResult> (plain.[EncodingPlainPrefix.Length..], options) |> nonNull
    | Some latex when latex.StartsWith EncodingLatexPrefix ->
        JsonSerializer.Deserialize<ExecutionResult> (latex.[EncodingLatexPrefix.Length..], options) |> nonNull
    | _ -> VoidSuccess

/// The cell that appends the last value to `run`, the history PreRunCode declares. `it` is
/// what F# Interactive binds the value of an expression cell to, so this is sent after every
/// cell that displayed a value, and it fails harmlessly -- and silently -- before `run` exists,
/// which is only during the warm-up cell.
/// https://github.com/asc-community/AngouriMath/issues/601
let private rememberCode = "run.Add (box it)"

let execute (kernel : FSharpKernel) code =
    let submitCode = SubmitCode code
    let computed = (kernel.SendAsync submitCode).Result    // Yes. It's Result.

    let mutable nonVoidResponse : string Option = None
    let mutable res : ExecutionResult Option = None

    for e in computed.Events do
        match e with
        | :? CommandSucceeded ->
            res <- objectDecode nonVoidResponse |> Some
        | :? CommandFailed as failed ->
            res <- Error failed.Message |> Some
        | :? DisplayEvent as display ->
            nonVoidResponse <- (Seq.head display.FormattedValues).Value |> Some
        | _ -> ()

    match res, nonVoidResponse with
    | Some (PlainTextSuccess _ | LatexSuccess _), Some _ when code <> rememberCode ->
        (kernel.SendAsync (SubmitCode rememberCode)).Result |> ignore
    | _ -> ()

    match res with
    | None -> EndOfFile
    | Some res -> res

let createKernel () =
    let kernel = new FSharpKernel ()

    let assemblyLoad = AssemblyLoadBuilder (execute, kernel)

    assemblyLoad {
        typeof<AngouriMath.MathS>
        typeof<AngouriMath.FSharp.Core.ParseException>
        typeof<AngouriMath.InteractiveExtension.KernelExtension>
        typeof<Plotly.NET.Chart>
        typeof<PeterO.Numbers.EDecimal>
        typeof<GenericTensor.Core.TensorShape>

        Formatter.SetPreferredMimeTypesFor(typeof<obj>, "text/plain")
        Formatter.Register<obj> objectEncode
        Formatter.Register<Entity.Matrix> (Func<Entity.Matrix, string> (fun m -> m.ToString(true) |> objectEncode), "text/plain")

        (fun c ->
            c 
            |> Chart.withSize (1200., 900.)
            |> Chart.show
            "Showing in the browser")
        |> (fun f -> Func<GenericChart, string> f)
        |> (fun f -> 
            Formatter.Register<GenericChart> (f, "text/plain"))
    }

File: Sources/Terminal/AngouriMath.Terminal.Shared/FSharpInteractive.cs
using Microsoft.DotNet.Interactive.Commands;
using Microsoft.DotNet.Interactive.Events;
using Microsoft.DotNet.Interactive.Server;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using static AngouriMath.Terminal.Shared.ExecutionResult;
using Microsoft.DotNet.Interactive.FSharp;
using System;
using Microsoft.DotNet.Interactive;
using Microsoft.DotNet.Interactive.Formatting;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Reactive.Linq;

namespace AngouriMath.Terminal.Shared
{
    public sealed class FSharpInteractive
    {
        private const string ENC_PLAIN_PREFIX = "encp";
        private const string ENC_LATEX_PREFIX = "encl";

        private static string ObjectEncode(object o)
            => o switch
            {
                Core.ILatexiseable iLatex =>
                    ENC_LATEX_PREFIX + JsonSerializer.Serialize(new LatexSuccess(iLatex.Latexise(), o.ToString() ?? "")),
                _ =>
                    ENC_PLAIN_PREFIX + JsonSerializer.Serialize(new PlainTextSuccess(o.ToString() ?? ""))
            };

        private static ExecutionResult ObjectDecode(string? inp)
            => inp switch
            {
                null => new VoidSuccess(),
                var plain when plain.StartsWith(ENC_PLAIN_PREFIX)
                    => JsonSerializer.Deserialize<PlainTextSuccess>(plain[ENC_PLAIN_PREFIX.Length..]) ?? throw new NullReferenceException(),
                var latex when latex.StartsWith(ENC_LATEX_PREFIX)
                    => JsonSerializer.Deserialize<LatexSuccess>(latex[ENC_LATEX_PREFIX.Length..]) ?? throw new NullReferenceException(),
                _ => new VoidSuccess()
            };

        private FSharpKernel kernel;

        private async Task<ExecutionResult> LoadAssembly(string path)
        {
            var loc = path.Replace(@"\", @"\\");
            return await Execute($"#r \"{loc}\"");
        }

        public static async Task<DUnion> Create()
        {
            var interactive = new FSharpInteractive();
            interactive.kernel = new FSharpKernel();
            
            if (await interactive.LoadAssembly(typeof(MathS).Assembly.Location) is Error err1)
                return DUnion.V(err1);
            if (await interactive.LoadAssembly(typeof(FSharp.Core).Assembly.Location) is Error err2)
                return DUnion.V(err2);
            if (await interactive.LoadAssembly(typeof(Interactive.AggressiveOperators).Assembly.Location) is Error err3)
                return DUnion.V(err3);

            Formatter.SetPreferredMimeTypeFor(typeof(object), "text/plain");
            Formatter.Register<object>(ObjectEncode);

            return DUnion.V(interactive);
        }

        private FSharpInteractive() { }

        public async Task<ExecutionResult> Execute(string code)
        {
            var submitCode = new SubmitCode(code);
            string? nonVoidResponse = null;
            ExecutionResult? res = null;
            var computed = await kernel.SendAsync(new SubmitCode(code));
            computed.KernelEvents.Subscribe(
                e =>
                {
                    switch (e)
                    {
                        case CommandSucceeded:
                            res = ObjectDecode(nonVoidResponse);
                            break;
                        case CommandFailed failed:
                            res = new Error(failed.Message);
                            break;
                        case DisplayEvent display:
                            nonVoidResponse = display.FormattedValues.First().Value;
                            break;
                    }
                });
            return res ?? new EOF();
        }
    }

    public abstract record ExecutionResult
    {
        public sealed record SuccessPackageAdded : ExecutionResult;
        public sealed record Error(string Message) : ExecutionResult;
        public sealed record VoidSuccess : ExecutionResult;
        public sealed record PlainTextSuccess(string Result) : ExecutionResult;
        public sealed record LatexSuccess(string Latex, string Source) : ExecutionResult;
        public sealed record EOF : ExecutionResult;
    }

    public abstract record DUnion
    {
        public static Item<T> V<T>(T Value) => new(Value);
        public sealed record Item<T>(T Value) : DUnion;
    }
}

namespace System.Runtime.CompilerServices
{
    internal static class IsExternalInit { }
}